src/site_apportionment/site_apportionment.py

```diff
@@ -113,86 +336,49 @@ def apportion_sites(df: pd.DataFrame, config: dict) -> pd.DataFrame:
-    value_cols = get_imputation_cols(config)
-
-    # Calculate the number of unique non-blank postcodes
-    df = count_unique_codes_in_col(df, postcode)
-
-    # Condition of long forms, many sites, instance >=1, non-null postcodes
-    cond = (df[form] == long_code) & (df[postcode + "_count"] > 1) & (df[ins] >= 1)
-
-    # Dataframe dfm with many products - for apportionment and Cartesian product
-    dfm = df.copy()[cond]
-    dfm = dfm.drop(columns=[postcode + "_count"], axis=1)
+    value_cols: List[str] = get_imputation_cols(config)
 
-    # Dataframe with everything else - save unchanged
-    df_out = df[~cond]
-    df_out = df_out.drop(columns=[postcode + "_count"], axis=1)
-
-    # df_codes: dataframe with codes and numerical values
-    group_cols = [ref, period]
-    code_cols = [product, civdef, pg_num]
-    df_codes = dfm.copy()[group_cols + code_cols + value_cols]
+    # Set short form percentages to 100
+    df = set_short_form_percentages(df)
 
-    # Remove blank products
-    df_codes = df_codes[df_codes[product].str.len() > 0]
+    # Calculate the number of unique non-blank postcodes
+    df = count_unique_postcodes_in_col(df)
 
-    # # De-duplicate by summation - possibly, not needed
-    df_codes = df_codes.groupby(group_cols + code_cols).agg(sum).reset_index()
+    # Split the dataframe into two based on whether there's more than one site (postcode)
+    multiple_sites_df, df_out = split_many_sites_df(df)
 
-    # df_sites: dataframe with postcodes, percents, and everyting else
-    site_cols = [x for x in df_cols if x not in (code_cols + value_cols)]
-    df_sites = dfm.copy()[site_cols]
+    # category_df: dataframe with codes and numerical values
+    category_df = create_category_df(multiple_sites_df, value_cols)
 
-    # Remove instances that have no postcodes
-    df_sites = df_sites[df_sites[postcode].str.len() > 0]
+    sites_df = create_sites_df(multiple_sites_df, orig_cols, value_cols)
 
     # Check for postcode duplicates for QA
-    df_sites["site_count"] = df_sites.groupby(group_cols + [postcode])[
-        postcode
-    ].transform("count")
-    df_duplicate_sites = df_sites[df_sites["site_count"] > 1]
-    num_duplicate_sites = df_duplicate_sites.shape[0]
-    if num_duplicate_sites:
-        SitesApportionmentLogger.info(
-            f"There are {num_duplicate_sites} duplicate sites."
-        )
+    count_duplicate_sites(sites_df)
 
     # Calculate weights
-    df_sites = weights(df_sites)
+    sites_df = calc_weights_for_sites(sites_df, groupby_cols)
 
     #  Merge codes to sites to create a Cartesian product
-    df_cart = df_sites.merge(df_codes, on=group_cols, how="inner")
+    df_cart = create_cartesian_product(sites_df, category_df)
 
     # Apply weights
-    for value_col in value_cols:
-        df_cart[value_col] = df_cart[value_col] * df_cart["site_weight"]
-
-    # Restore the original column order
-    df_cart = df_cart[df_cols]
```

Comment:
I am not sure but I would keep it, just in case. I don't want to append two dataframes with different order of columns. Also, this will take care of additional columns in df_cart, if we forgot to drop them.